File: 05_arrays/05_20_majority_element_II/majority_element_II.cpp
#include<bits/stdc++.h>

using namespace std;

vector<int> majorityElement(vector<int>& nums) {
    vector<int> result;
    unordered_map<int, int> mp;
    int size = nums.size();
    int threshold = size / 3;

    for (int i = 0; i < size; i++)
    {
        mp[nums[i]]++;
    }
    

    for (auto pair : mp) {
        int element = pair.first;
        int count = pair.second;

        if (count > threshold)
        {
            result.push_back(element);
        }
    }

    return result;
}

int main()
{
    vector<int> arr = {11, 33, 33, 5, 33, 11};
    vector<int> ans = majorityElement(arr);
    cout << "The majority elements are: ";
    for (auto it : ans)
        cout << it << " ";
    cout << "\n";
    return 0;
}

File: 05_arrays/05_20_majority_element_II/majority_element_II.js
const majorityElement = (arr = []) => {
    const result = [];
    const map = new Map();
    const threshold = Math.floor(arr.length / 3);

    for (let i = 0; i < arr.length; i++) {
        map.set(arr[i], ((map.get(arr[i]) || 0) + 1));
    }

    console.log(map);

    for (const [key, value] of map) {
        if (value > threshold) {
            result.push(key);
        }
    }

    return result;
}


console.log(majorityElement([11, 33, 22, 11, 11, 33, 33]));